library/std/src/sys/uefi/net/tcp4.rs

```diff
@@ -0,0 +1,473 @@
+        let r = connection_token.completion_token.status;
+        if r.is_error() { Err(status_to_io_error(r)) } else { Ok(()) }
+    }
+
+    pub(crate) fn transmit(&self, buf: &[u8]) -> io::Result<usize> {
+        let buf_size = buf.len() as u32;
+        let transmit_event = common::Event::create(
+            r_efi::efi::EVT_NOTIFY_WAIT,
+            r_efi::efi::TPL_CALLBACK,
+            Some(nop_notify4),
+            None,
+        )?;
+        let completion_token =
+            tcp4::CompletionToken { event: transmit_event.as_raw_event(), status: Status::ABORTED };
+        let fragment_table = tcp4::FragmentData {
+            fragment_length: buf_size,
+            // FIXME: Probably dangerous
+            fragment_buffer: buf.as_ptr() as *mut crate::ffi::c_void,
+        };
+
+        let layout = unsafe {
+            crate::alloc::Layout::from_size_align_unchecked(
+                crate::mem::size_of::<tcp4::TransmitData>()
+                    + crate::mem::size_of::<tcp4::FragmentData>(),
+                POOL_ALIGNMENT,
+            )
+        };
+        let mut transmit_data = VariableBox::<tcp4::TransmitData>::try_new_uninit(layout)
+            .map_err(|_| common::VARIABLE_BOX_ERROR)?;
+
+        // Initialize TransmitData
+        unsafe {
+            let ptr = transmit_data.as_mut_ptr();
+            addr_of_mut!((*ptr).push).write(r_efi::efi::Boolean::TRUE);
+            addr_of_mut!((*ptr).urgent).write(r_efi::efi::Boolean::FALSE);
+            addr_of_mut!((*ptr).data_length).write(buf_size);
+            addr_of_mut!((*ptr).fragment_count).write(1);
+            (*ptr).fragment_table.as_mut_ptr().copy_from([fragment_table].as_ptr(), 1);
+        };
+        let mut transmit_data = unsafe { transmit_data.assume_init() };
+
+        let packet = tcp4::IoTokenPacket { tx_data: transmit_data.as_mut() };
+        let mut transmit_token = tcp4::IoToken { completion_token, packet };
+        unsafe { Self::transmit_raw(self.protocol.as_ptr(), &mut transmit_token) }?;
+
+        transmit_event.wait()?;
+
+        let r = transmit_token.completion_token.status;
+        if r.is_error() {
+            Err(status_to_io_error(r))
+        } else {
+            Ok(unsafe { (*transmit_token.packet.tx_data).data_length } as usize)
+        }
+    }
+
+    pub(crate) fn transmit_vectored(&self, buf: &[IoSlice<'_>]) -> io::Result<usize> {
+        let buf_size = crate::mem::size_of_val(buf);
+        let transmit_event = common::Event::create(
+            r_efi::efi::EVT_NOTIFY_WAIT,
+            r_efi::efi::TPL_CALLBACK,
+            Some(nop_notify4),
+            None,
+        )?;
+        let completion_token =
+            tcp4::CompletionToken { event: transmit_event.as_raw_event(), status: Status::ABORTED };
+        let fragment_tables: Vec<tcp4::FragmentData> = buf
+            .iter()
+            .map(|b| tcp4::FragmentData {
+                fragment_length: crate::mem::size_of_val(b) as u32,
```

Comment:
This is not correct: sizes larger than `u32::MAX` will be truncated down, so the buffers will not be fully transmitted. Use a custom `IoSlice` to resolve this (possibly by copying `sys::windows::io`). Also, `b.len()` works, no need for `size_of_val` here.